### PR TITLE
issue: 1491767 Use netlink api for traffic control

### PR DESCRIPTION
**Pull request by igor-ivanov**

### CODE REVIEW

**Comment by liranoz12** [comment]
bot:retest

File: tools/daemon/tc.c [hunk 1]
```diff
@@ -0,0 +1,812 @@
...
+	if (ip) {
+		pack_key32(&opt_sel.sel, ntohl(ip), 0xffffffff, 16, 0);
+	} else {
+		pack_key32(&opt_sel.sel, ntohl(ip), 0, 0, 0);
+	}
+	opt_sel.sel.flags |= TC_U32_TERMINAL;
+	nl_attr_add(&tc->req.hdr, TCA_U32_SEL, &opt_sel, sizeof(opt_sel.sel) + opt_sel.sel.nkeys * sizeof(opt_sel.sel.keys[0]));
+
+	nl_attr_nest_end(&tc->req.hdr, opts);
+
```
Comment (liranoz12) [inline tools/daemon/tc.c]: We need to talk about it...

File: tools/daemon/tc.c [hunk 2]
```diff
@@ -0,0 +1,812 @@
...
+	if (nl_send(tc->nl, &tc->req.hdr) < 0) {
+		rc = -1;
+		goto err;
+	}
+	if (nl_recv(tc->nl, NULL, NULL) < 0) {
+		rc = -1;
+		goto err;
+	}
```
Comment (liranoz12) [inline tools/daemon/tc.c]: We need to talk about it...

File: tools/daemon/tc.c [hunk 3]
```diff
@@ -0,0 +1,812 @@
...
```
Comment (liranoz12) [inline tools/daemon/tc.c]: Why do we need to function for 3t/5t ?
Can we use if statement inside ?

File: tools/daemon/tc.c [hunk 4]
```diff
@@ -0,0 +1,812 @@
...
+	struct rtattr *opts = NULL;
+	struct {
+		struct tc_u32_sel sel;
+		struct tc_u32_key keys[10];
+	} opt_sel;
+
+	tc_req(tc, ifindex, RTM_NEWTFILTER,
+			(NLM_F_REQUEST | NLM_F_ACK | NLM_F_EXCL | NLM_F_CREATE),
+			qdisc);
+
```
Comment (liranoz12) [inline tools/daemon/tc.c]: This section is duplicate by 4 function. please try to use additional function in order to avoid this duplication

Comment (igor-ivanov) [reply tools/daemon/tc.c]: there are differences

File: tools/daemon/tc.c [hunk 5]
```diff
@@ -0,0 +1,812 @@
...
+	tc->req.hdr.nlmsg_type = type;
+	tc->req.hdr.nlmsg_flags = (flags ? flags : (NLM_F_REQUEST | NLM_F_ACK));
+	tc->req.hdr.nlmsg_pid = 0; /* to communicate kernel */
+	tc->req.hdr.nlmsg_seq = 0; /* update during send */
+
+	tc->req.msg.tcm_family = AF_UNSPEC;
+	tc->req.msg.tcm_ifindex = ifindex;
+	tc->req.msg.tcm_handle = qdisc.handle;
+	tc->req.msg.tcm_parent = qdisc.parent;
+	tc->req.msg.tcm_info = TC_H_MAKE(qdisc.prio << 16, htons(ETH_P_IP));
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply tools/daemon/tc.c]: tcm_info consists of {prio,proto} values

File: tools/daemon/tc.c [hunk 6]
```diff
@@ -0,0 +1,812 @@
...
+
+	tc_req(tc, ifindex, RTM_DELQDISC,
+			0,
+			qdisc);
+
+	nl_attr_add(&tc->req.hdr, TCA_KIND, "ingress", sizeof("ingress"));
+
+	opts = nl_attr_nest_start(&tc->req.hdr, TCA_OPTIONS);
+	nl_attr_nest_end(&tc->req.hdr, opts);
+
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply tools/daemon/tc.c]: dpdk does not use tc

File: tools/daemon/nl.c [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+	}
+	nl_msg->nlmsg_len = NLMSG_ALIGN(nl_msg->nlmsg_len) + RTA_ALIGN(rta->rta_len);
+}
+
+struct rtattr *nl_attr_nest_start(struct nlmsghdr *nl_msg, int type)
+{
+	struct rtattr *nest = NLMSG_TAIL(nl_msg);
+
+	nl_attr_add(nl_msg, type, NULL, 0);
+
```
Comment (liranoz12) [inline tools/daemon/nl.c]: need to verify that no need to malloc new element here (and free in nl_attr_nest_end)

File: tools/daemon/tc.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (liranoz12) [reply tools/daemon/tc.c]: they do  : 
https://git.dpdk.org/dpdk/tree/drivers/net/tap/tap_tcmsgs.c#n73

File: tools/daemon/tc.c [hunk 8]
```diff
@@ -0,0 +1,812 @@
...
+	tc_req(tc, ifindex, RTM_NEWQDISC,
+			(NLM_F_REQUEST | NLM_F_ACK | NLM_F_EXCL | NLM_F_CREATE),
+			qdisc);
+
+	nl_attr_add(&tc->req.hdr, TCA_KIND, "ingress", sizeof("ingress"));
+
+	opts = nl_attr_nest_start(&tc->req.hdr, TCA_OPTIONS);
+	nl_attr_nest_end(&tc->req.hdr, opts);
+
```
[the remark this answers is not in the document]
Comment (liranoz12) [reply tools/daemon/tc.c]: see https://git.dpdk.org/dpdk/tree/drivers/net/tap/tap_tcmsgs.c#n147

File: tools/daemon/nl.c [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+	msg.msg_namelen = sizeof(nladdr);
+	msg.msg_iov = &iov;
+	msg.msg_iovlen = 1;
+
+	do {
+		struct nlmsghdr *nl_msg;
+		int recv_bytes = 0;
+
+		recv_bytes = recvmsg(nt->fd, &msg, 0);
+		if (recv_bytes <= 0) {
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply tools/daemon/nl.c]: We expect getting answer in all cases. As a result no answer or performed an orderly shutdown that are equal to return 0 means failure for as.

File: tools/daemon/nl.c [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+ * @struct nl_object
+ * @brief netlink container
+ */
+struct nl_object {
+	int fd;            /**< the netlink socket file descriptor used for communication */
+	int seq;           /**< sequence number of send operation */
+	char buf[81920];   /**< buffer for receive data */
+};
+
+nl_t nl_create(void)
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply tools/daemon/nl.c]: moved to daemon_cfg

File: tools/daemon/nl.c [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+		memset(nt, 0, sizeof(*nt));
+		nt->fd = fd;
+		nt->seq = 0;
+	}
+
+	return nt;
+err:
+	free(nt);
+	nt = NULL;
+
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply tools/daemon/nl.c]: no need

File: tools/daemon/flow.c [hunk 12]
```diff
@@ -110,23 +95,35 @@ int del_flow(struct store_pid *pid_value, struct store_flow *value);
...
-static inline void add_pending_list(pid_t pid, struct flow_ctx *ctx, char* if_name, int ht_id, int prio, int *rc);
-static inline void free_pending_list(pid_t pid, struct flow_ctx *ctx, char* if_name);
+static inline void add_pending_list(pid_t pid, struct flow_ctx *ctx, int if_index, int ht_id, int prio, int *rc);
+static inline void free_pending_list(pid_t pid, struct flow_ctx *ctx, int if_index);
 static inline int get_prio(struct store_flow *value);
 static inline int get_bkt(struct store_flow *value);
 static inline int get_protocol(struct store_flow *value);
 static inline int get_node(struct store_flow *value, struct list_head **list);
 
+static tc_t tc = NULL;
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply tools/daemon/flow.c]: moved to daemon_cfg

File: tools/daemon/daemon.h [hunk 13]
```diff
@@ -361,4 +373,50 @@ static inline int sys_iplocal(uint32_t addr)
...
+		return;
+		out_pos += ret;
+		for (j = 0; j < 16; j++) {
+			if (i + j < buflen)
+			ret = sprintf(out_buf + out_pos, "%02x ",
+					buf[i + j]);
+			else
+			ret = sprintf(out_buf + out_pos, "   ");
+			if (ret < 0)
+			return;
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply tools/daemon/daemon.h]: 👍

File: tools/daemon/daemon.h [hunk 14]
```diff
@@ -67,6 +68,10 @@
 #define EXIT_SUCCESS 0
 #define EXIT_FAILURE 1
 
+#ifndef UNREFERENCED_PARAMETER
+#define UNREFERENCED_PARAMETER(P) ((void)P)
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply tools/daemon/daemon.h]: 👍

File: tools/daemon/nl.c (hunk 7 above)
Comment (igor-ivanov) [reply tools/daemon/nl.c]: no need because all are done on stack

Comment (igor-ivanov) [reply tools/daemon/nl.c]: it is done on stack in current implementation

File: tools/daemon/tc.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply tools/daemon/tc.c]: we meet original tc utility format

**Comment by igor-ivanov** [comment]
updated with review changes

**Pull request merged**